### PR TITLE
MF-293: Buttons on the home page should be extensions

### PR DESCRIPTION
**Pull request by denniskigen**
https://issues.openmrs.org/browse/MF-293

This PR:
- Creates an extension slot on the home page called `home-page-buttons`,
- Attaches an extension called `registration-link` to that extension slot - this extension is a link to the registration page that gets styled as a button and then is bound to the extension slot.
- Removes the registration button config from the home ESM config in `root.component.tsx` as that is now taken care of via the extension system.

Kapture 2020-09-09 at 22 18 24

### CODE REVIEW

**Comment by jonathandick** [comment]
@denniskigen could you leave a gif that shows what's going on with this extension?

**Comment by denniskigen** [comment]
@jonathandick yeah, I've updated the PR description. This work is not complete - there's some stuff to do with dependencies that needs sorting out e.g. publishing the extension manager package.

File: src/dashboard/home-dashboard.component.tsx [hunk 1]
```diff
@@ -27,7 +31,14 @@ export default function HomeDashboard(props: HomeDashboardProps) {
               </Link>
             </div>
             {config.buttons.enabled && (
-              <div className={styles.buttonArea}>{buttons}</div>
+              <div className={styles.buttonArea}>
+                <ExtensionSlotReact extensionSlotName="home-page-buttons">
```
Comment (jonathandick) [inline src/dashboard/home-dashboard.component.tsx:35]: given that the name is plural "home-page-button**s**", do you expect multiple extensions to attach here? If there are multiple extensions, what determines the ordering?

Comment (denniskigen) [reply src/dashboard/home-dashboard.component.tsx:35]: Yes I do, an arbitrary number of extensions can be attached to an extension slot. The order is determined by the order in which the extensions are passed in the call to `attach`.

**Comment by brandones** [comment]
Great progress, Dennis.

So here's the tricky thing—we want this extension to be provided by esm-registration-app. That means that it should be defined in esm-registration-app as a completely plain
```jsx
    <a
      href={url}
      onClick={event => singleSpaNavigate(event, url)}
    >
      Patient Registration
    </a>
```

that is, with no styling. You could imagine this extension also being added as a button in the primary navbar, or wherever else—it's just a link.

The job of styling it then belongs to esm-home. It'll do something like
```jsx
                <ExtensionSlotReact extensionSlotName="home-page-buttons">
                  <div className={style.homeButton}>
                    <ExtensionReact />
                  </div>
                </ExtensionSlotReact>
```

with

```css
.homeButton a {
  /* whatever styling needs to be applied to that link */
}
```

Hope that makes sense. Forgive me if this is already the track you were on, and you were just at an intermediate stage with the draft PR.

**Comment by denniskigen** [comment]
> Great progress, Dennis.
> 
> So here's the tricky thing—we want this extension to be provided by esm-registration-app. That means that it should be defined in esm-registration-app as a completely plain
> 
> ```js-jsx
>     <a
>       href={url}
>       onClick={event => singleSpaNavigate(event, url)}
>     >
>       Patient Registration
>     </a>
> ```
> 
> that is, with no styling. You could imagine this extension also being added as a button in the primary navbar, or wherever else—it's just a link.
> 
> The job of styling it then belongs to esm-home. It'll do something like
> 
> ```js-jsx
>                 <ExtensionSlotReact extensionSlotName="home-page-buttons">
>                   <div className={style.homeButton}>
>                     <ExtensionReact />
>                   </div>
>                 </ExtensionSlotReact>
> ```
> 
> with
> 
> ```css
> .homeButton a {
>   /* whatever styling needs to be applied to that link */
> }
> ```
> 
> Hope that makes sense. Forgive me if this is already the track you were on, and you were just at an intermediate stage with the draft PR.

Not at all @brandones , this was very useful. I've updated the PR to match this and opened the related PR in `opeenmrs-esm-patient-registration`.

**Review by brandones — APPROVED** [review]
Looks great, thanks @denniskigen ! I released esm-extension-manager, so the latest version of that should work fine.

**Comment by brandones** [comment]
Just make sure to get those package deps fixed and the build working before merging :)

**Review by brandones — APPROVED** [review]

**Pull request merged**